### PR TITLE
refactor: Make kubernetes in runtime

### PR DESCRIPTION
**Pull request by gaocegege**
This PR is to abstract the access to Kubernetes, and make it a runtime interface, instead of using it directly

In the future, we may support remote machine as a new runtime.

Signed-off-by: Ce Gao <cegao@tensorchord.ai>

### CODE REVIEW

**Review by VoVAllen — APPROVED** [review]
lgtm

**Review by kemingy — DISMISSED** [review]
/lgtm

File: pkg/runtime/kubernetes/util.go [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+			return nil, errors.Wrap(err, "failed to parse memory resource")
+		}
+		resRequest[v1.ResourceMemory] = mem
+	}
+	if res.GPU != "" {
+		gpu, err := resource.ParseQuantity(res.GPU)
+		if err != nil {
+			return nil, errors.Wrap(err, "failed to parse gpu resource")
+		}
+		resRequest["nvidia/gpu"] = gpu
```
Comment (xieydd) [inline pkg/runtime/kubernetes/util.go:41]: How about creating a const.go to store `nvidia/gpu`.

File: pkg/runtime/kubernetes/environment_get.go [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+	}
+	if pod.Labels[consts.PodLabelUID] != owner {
+		logrus.WithFields(logrus.Fields{
+			"loginname_in_pod":     pod.Labels[consts.PodLabelUID],
+			"loginname_in_request": owner,
+		}).Debug("mismatch loginname")
+		return nil, errdefs.Unauthorized(errors.New("mismatch loginname"))
+	}
+
+	if pod == nil {
```
Comment (xieydd) [inline pkg/runtime/kubernetes/environment_get.go:39]: maybe duplicate, because have already check in `k8serrors.IsNotFound(err)`.

**Review by xieydd — COMMENTED** [review]
some nits.

File: pkg/runtime/kubernetes/util.go [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+		Status: types.EnvironmentStatus{},
+	}
+
+	// Convert the spec.
+	if len(pod.Spec.Containers) != 0 {
+		env.Spec.Image = pod.Spec.Containers[0].Image
+		env.Spec.Env = envVarsfromKubernetes(pod.Spec.Containers[0].Env)
+		env.Spec.Cmd = pod.Spec.Containers[0].Command
+	}
+	if pod.Labels[consts.PodLabelUID] == "" {
```
[the remark this answers is not in the document]
Comment (gaocegege) [reply pkg/runtime/kubernetes/util.go:66]: Thanks for the catch!

File: pkg/runtime/kubernetes/util.go (hunk 1 above)
Comment (gaocegege) [reply pkg/runtime/kubernetes/util.go:41]: Could you please explain more about it?

Comment (xieydd) [reply pkg/runtime/kubernetes/util.go:41]: Just like `PodLabelUID`, maybe useful in other file.

Comment (gaocegege) [reply pkg/runtime/kubernetes/util.go:41]: Do you mean to make a variable to store the string?

Comment (xieydd) [reply pkg/runtime/kubernetes/util.go:41]: right.

**Comment by gaocegege** [comment]
Comments are addressed, please take a look.
Thanks for your time :smile:

**Review by xieydd — APPROVED** [review]
LGTM

**Pull request merged**